moped-editor/src/views/projects/newProjectView/NewProjectMap.js

```diff
@@ -203,6 +247,81 @@ const NewProjectMap = ({
     saveActionDispatch
   );
 
+  /**
+   * Handles when a new map feature is selected and updates feature collection state
+   * @param {Object} selectedFeature - The selected feature geojson
+   * @return {undefined} Side effect of calling setFeatureCollection with new state
+   */
+  const handleSelectedFeatureUpdate = selectedFeature => {
+    const isPresent = isFeaturePresent(
+      selectedFeature,
+      featureCollection.features,
+      selectedFeature.properties.sourceLayer
+    );
+
+    if (selectedFeature.properties.sourceLayer !== "CTN" || isPresent) {
+      // we are only concerned with CTN (aka line) features that are being added to the feature collection
+      // otherwise handle normally
+      commitFeatureCollectionUpdate(
+        selectedFeature,
+        featureCollection,
+        setFeatureCollection,
+        isPresent
+      );
+      return;
+    }
+
+    // we must decide if the line the feature may be fragmented from tiling
+    const map = mapRef.current.getMap();
+    // get the current map bounds (LngLatBounds instance) and convert it to a simple array
+    const bbox = map
+      .getBounds()
+      .toArray()
+      .flat();
+    // convert the bbox array to a geojson line
+    const bboxLine = polygonToLine(bboxPolygon(bbox));
+    const intersectsWithBounds = booleanIntersects(bboxLine, selectedFeature);
+
+    if (!intersectsWithBounds) {
+      // feature is entirely contained—we have it's complete geometry
+      commitFeatureCollectionUpdate(
+        selectedFeature,
+        featureCollection,
+        setFeatureCollection,
+        isPresent
+      );
+      return;
+    }
+    // this feature is rendered to the edge of the viewport and may be fragmented. we cannot
+    // know with certainty, so we fetch it's complete geometry from AGOL to be safe.
+    // also considered turf.booleanWithin() and turf.booleanContains - not reliable.
+    const selectedFeatureId = getFeatureId(
+      selectedFeature,
+      selectedFeature.properties.sourceLayer
+    );
+
+    queryCtnFeatureService(selectedFeatureId).then(queriedFeatureCollection => {
+      // Update the selectedFeature geometry if a feature has been found. To potential
+      // error cases are handled here:
+      //    1. fetch error: queriedFeatureCollection is null
+      //    2. feature is not found (queriedFeatureCollection.features is empty)
+      // In both cases we simply use the geometry of the selectedFeature which may be
+      // fragemented but is better than nothing.
+      selectedFeature.geometry =
+        queriedFeatureCollection?.features?.[0]?.geometry ||
+        selectedFeature.geometry;
+
+      commitFeatureCollectionUpdate(
+        selectedFeature,
+        featureCollection,
+        setFeatureCollection,
+        isPresent
+      );
+    });
+
+    return;
+  };
+
```

Comment:
Absolutely brilliant, I personally did not experience lags and it feels seamless, but I can see this being an issue if the user's computer is slow or the internet connection is bad.

Reply:
@johnclary out of curiosity, how large is the size of the full CTM map?